src/core/io/platform/f4xx/CANf4xx.cpp

```diff
@@ -0,0 +1,346 @@
+
+
+#include <cstring>
+#include <stdint.h>
+
+#include <HALf4/stm32f4xx.h>
+
+#include <core/io/pin.hpp>
+#include <core/io/platform/f4xx/CANf4xx.hpp>
+#include <core/io/platform/f4xx/GPIOf4xx.hpp>
+#include <core/platform/f4xx/stm32f4xx.hpp>
+#include <core/utils/log.hpp>
+#include <core/utils/time.hpp>
+#include <core/utils/types/FixedQueue.hpp>
+
+namespace {
+
+/**
+ * Struct to hold the global pointers to the CAN F4 interfaces. This is necessary
+ * for the interrupts to have a call back to the classes.
+ */
+struct CANf4Global {
+    // Pointer to the CANf446xx interface
+    core::io::CANf4xx* canInstance = nullptr;
+
+    // Pointer to the CAN handler for HAL,
+    CAN_HandleTypeDef* canHandler = nullptr;
+};
+
+struct CANf4Global globalCAN[2];
+
+// Interrupt handler for CAN 1 or index 0 in the global CAN array
+extern "C" void CAN1_RX0_IRQHandler(void) {
+    if (globalCAN[0].canHandler != nullptr) {
+        HAL_CAN_IRQHandler(globalCAN[0].canHandler);
+    }
+}
+
+// Interrupt handler for CAN 2 or index 1 in the global CAN array
+extern "C" void CAN2_RX0_IRQHandler(void) {
+    if (globalCAN[1].canHandler != nullptr) {
+        HAL_CAN_IRQHandler(globalCAN[1].canHandler);
+    }
+}
+
+/**
+ * HAL interrupt handler for incoming CAN messages. The messages are added to
+ * a queue for receiving.
+ */
+extern "C" void HAL_CAN_RxFifo0MsgPendingCallback(CAN_HandleTypeDef* hcan) {
+    CAN_RxHeaderTypeDef rxHeader = {0};
+    uint8_t payload[core::io::CANMessage::CAN_MAX_PAYLOAD_SIZE];
+
+    HAL_CAN_GetRxMessage(hcan, CAN_RX_FIFO0, &rxHeader, payload);
+
+    // Construct the CANmessage
+    bool isExtended = rxHeader.IDE == CAN_ID_EXT;
+    uint32_t id     = isExtended ? rxHeader.ExtId : rxHeader.StdId;
+    core::io::CANMessage message(id, rxHeader.DLC, payload, isExtended);
+
+    core::io::CANf4xx* canInstance;
+    if (hcan == globalCAN[0].canHandler) {
```

Comment:
That's fine since the pointer is bound to the physical hardware peripheral address and will never change for the chip. Also this is some of the BS the HAL does.